### ISSUE TITLE
Add ConnectivityState.TERMINATED or change SHUTDOWN to notify after gRPC threads end.

### ISSUE THREAD

**Issue by anuraaga** [grpc/grpc-java#8432]
### Is your feature request related to a problem?

We shutdown `ManagedChannel` asynchronously so need an asynchronous callback for determining when a channel has been completely cleaned up. This can be important for when libraries are dynamically unloaded / loaded at runtime as the order matters. Currently there is only `ConnectivityState.SHUTDOWN`, which means no new RPCs will be serviced, but the connection may still be operating. Notably, there is almost always a `GOAWAY` frame sent after this connectivity state is reached.

### Describe the solution you'd like

Add `ConnectivityState.TERMINATED` to notify after connections have been completely closed. This would be sent after all gRPC threads have been ended.

### Describe alternatives you've considered

`ConnectivityState.SHUTDOWN` could possibly be changed to report after gRPC threads have ended - it's unclear to me what the use case is of being notified of shutdown before this has happened but there may be a use case, in which case a new enum is needed instead. That being said, because it's only a timing issue, it would be rare to break applications by changing the behavior this way too and could be viable.

**Comment by sanjaypujare** [comment]
> ConnectivityState.SHUTDOWN could possibly be changed to report after gRPC threads have ended 

That cannot be done since "shutdown" and "terminated" have clear meanings (in this context). See [`isShutdown`](https://github.com/grpc/grpc-java/blob/master/api/src/main/java/io/grpc/ManagedChannel.java#L44) and [`isTerminated`](https://github.com/grpc/grpc-java/blob/master/api/src/main/java/io/grpc/ManagedChannel.java#L53).

So the only possibility is to add `ConnectivityState.TERMINATED` as you suggest. So the channel state becomes `ConnectivityState.TERMINATED` when `isTerminated()` is `true` for that channel. This will need changes in [`ConnectivityStateManager.gotoState()`](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ConnectivityStateManager.java#L62). Do you want to submit a PR?

Note that there is a [`ManagedClientTransport.Listener.transportTerminated()`](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ManagedClientTransport.java#L92) callback, but this is per subchannel and a channel has one or more subchannels.

**Comment by ejona86** [comment]
The channel state API uses the Channel's executor to run callbacks, but after termination the Channel's executor may be shutdown/unusable (generally it won't but we shouldn't make an API requiring it to be available). Other than all the pain of changing the channel state API.

For the opentelemetry-java use-case, is the problem that RPCs are still on the transport, so it may take a while for the channel to terminate?

NoClassDefFoundError for being "unloaded" by a ClassLoader sounds like a dubious ClassLoader. That's not how they should work. The problem seems to be the ClassLoader is no longer loading more classes and the shutdown code path uses an anonymous class. Note that with this dubious ClassLoader behavior, waiting for termination [is not enough](https://github.com/grpc/grpc-java/issues/8309).

The [current opentelemetry-java code](https://github.com/open-telemetry/opentelemetry-java/blob/86e7d127e26b5a4a1a48301bf5e0d775c010042d/exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/OtlpGrpcSpanExporter.java#L181-L185) is not really doing much useful with the `notifyWhenStateChanged`. The Channel enters the channel state [as the _first thing_](https://github.com/grpc/grpc-java/blob/e45aab085c45eed46c94cd705e2e53cdf69a3f29/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L814) when calling shutdown(). That is to say the channel state API 1) was changed to SHUTDOWN before traffic began to be rejected and 2) is highly likely to be SHUTDOWN when `shutdown()` returns (the SynchronizationContext means some other thread may actually complete the transition).

But actually, `managedChannel.notifyWhenStateChanged(ConnectivityState.SHUTDOWN, result::succeed);` is executing the callback _before_ `shutdown()` is called. The state passed to channel state API is the _current_ state and the callback will be notified when the state _no longer matches_ the provided state. Thus it is asking for a callback when the channel _isn't shut down._

Since this callback is needed to be called _after_ termination, we will need an API where the caller provides an Executor. Looking at the code, it seems there isn't any thread available, so it'd likely use `directExecutor()` which would likely mean running on a Netty thread and then run any `whenComplete()`s directly on the Netty thread as well.

Generally I've not seen good things from asynchronous termination APIs, because the thread you use to notify of the termination is supposed to be terminated. From a technical standpoint I think the API would be easy to implement, but "does it do something useful" and not open a can of worms seems up for debate.

**Comment by anuraaga** [comment]
> Thus it is asking for a callback when the channel isn't shut down.

Yeah I noticed the mistake while looking through the code and do see the javadoc describe it. I'm used to APIs to register a callback when something is true, but here it's when something isn't true - I guess I didn't notice this earlier and does feel somewhat unergonomic.

> From a technical standpoint I think the API would be easy to implement, but "does it do something useful" and not open a can of worms seems up for debate.

I admittedly did use the ending of a thread in the issue description but agree that notification of thread end is mostly impossible. It seems true for the blocking version of the API too and not unique to async - the block just ensures a signal was received from the live thread but not that the thread has already finished.

The key point is knowing no more I/O will be done - it seems useful to present this information. Currently it's only through the blocking awaitTermination but if an async handler could be invoked similarly, it would fit with our api. The main reason the Otel results are  async is to make sure not to limit user cases - it's very easy to block on a result by calling join but very hard to treat a blocking call asynchronously by spawning a thread. It's the situation we're currently in on this issue ;) We also want to make the API consistent - any method can be called on an event loop as needed, no need to worry about potential blocking anywhere, even shutdown of a client stub, if that's important for the app.

**Comment by ejona86** [comment]
> From a technical standpoint I think the API would be easy to implement

Hmmm... We are deep in the synchronization context when the channel become terminated. I'd rather not be forced to issue callbacks while in the synchronization context, doubly so since callers (and specifically you) are so likely to use directExecutor(). If we could get out of the synchronization context easily, I'd probably be more okay with it... Maybe. I'm really on the fence. We have tried really hard in gRPC to avoid applications running on our core threads, and it has been a huge blessing.

> The main reason the Otel results are async is to make sure not to limit user cases

FWIW, that sort of argument works better if you are paying the resources to make it happen. "Someone might want it" is a very weak argument when having a concrete discussion. I do agree that "everything being async" can be very nice, but someone has to "pay" for the thread in cases like this, so generally the caller has to pay that thread so they can potentially share resources across such cases. But here it seems nobody has a thread to provide.

**Comment by anuraaga** [comment]
> We have tried really hard in gRPC to avoid applications running on our core threads, and it has been a huge blessing.

Is this different from the `directExecutor` flag on the channel or server builders?

https://github.com/grpc/grpc-java/blob/master/api/src/main/java/io/grpc/ManagedChannelBuilder.java#L93

I figured that running callbacks on gRPC threads, while not the default, is a supported model. Though perhaps the difference is transport thread vs synchronization thread which is admittedly more of a

If the possibility of being passed in `directExecutor` is off the table, then there probably isn't an approach here. FWIW, I think with asynchronous APIs this is fairly standard, ListenableFuture and CompletableFuture both allow setting an executor but that could be direct - it's up to the caller to know their callback is safe to run directly on an unknown thread, usually by it being a small amount of logic like a couple of lines signaling another future, or otherwise just don't do it. There could be some concern about the possibility of foot-shooting, I'd say the pattern is idiomiatic in Java 8+.
